### ISSUE TITLE
怎么扫描不到IP了？

### ISSUE THREAD

**Issue by Asteroids89** [Kisesy/gscan_quic#3]
主任，自己编译的最新版的gscan_quic扫描IP一直是0，旧版的还有一些呢，是现在封的太厉害，还是怎么回事？
![qq 20170715230027](https://user-images.githubusercontent.com/19388220/28240301-c1e99344-69b1-11e7-8377-0c26489ea2f9.png)

**Comment by Kisesy** [comment]
![image](https://user-images.githubusercontent.com/5843694/28240352-f2717292-69b2-11e7-94d9-1f8fb6c6b13c.png)

不好重现，待会看看

**Comment by Kisesy** [comment]
哎？我发现自行编译还不是编译的最新版

**Comment by jap33** [comment]
xx里面那么多ip段。

**Comment by Kisesy** [comment]
感觉很怪，最新代码我自己编译的就可以把IP段打乱顺序扫描，但是网页编译的却不可以
不理解什么地方出了问题

**Comment by MeABc** [comment]
....网页编译还有缓存？

**Comment by MeABc** [comment]
Travis CI Build

**Comment by Kisesy** [comment]
@MeABc 真有，刚刚我测试新代码，发现新编译慢多了...新代码已经可以打乱顺序了

**Comment by Kisesy** [comment]
下载 https://gox.jpillora.com/ 最新版，里面 1.1的那个试试

**Comment by Kisesy** [comment]
@MeABc Travis CI Build 编译还是挺麻烦的，其实一开始，我以为这个工具一般也用不着更新的 :trollface:

**Comment by MeABc** [comment]
哈哈，情况就是这样子的 --- 一个人用的话哪有这么问题（很难发现），变成公开项目就是这样。。。

**Comment by MeABc** [comment]
不过这样也好，我以后常来 ---- 过来抄代码。。。。哈哈。。。

**Comment by MeABc** [comment]
https://github.com/johnsonz/go-checkiptools/issues/9

这是我 gvs/gws 测试实现的方法。之前和 SeaHOH 讨论过的。简单说就是在 TLS 握手阶段完成所有的判断。供你参考 ---- 如果你打算继续修改 gscan 的话。

**Comment by wpyok500** [comment]
楼主的问题我在公司使用也出现过，但家里使用就没出现，可能跟网咯有关。

**Comment by Kisesy** [comment]
@MeABc 当然，这是最简单的，也是最快的
其实 @SeaHOH 在之前的帖子里也说过不要用 client，但是这个 quic 的情况比较复杂
其实用 client 是算简单的了，首先来说quic用的是udp，即使是个无效地址，也会连接成功
所以要验证就要tls握手，光握手还不行，一般的https还行，会握手失败，但是一些cdn也会握成功，这时候还是得真正的http来测
一开始我以为 quic 只是一种传输格式，里面该传什么传什么，比如可以往里写 GET / HTTP 之类的，但是它只会返回 0 值。。。

**Comment by MeABc** [comment]
@Kisesy 

> 但是这个 quic 的情况比较复杂

是的。尤其是 quic-go 的代码，我都不知道从哪下手 ---- 只从握手阶段来做判断 --- quic 的协议现在还没有真正搞明白。

**Comment by SeaHOH** [comment]
如果不判断证书，应该可以实现。
只发送协商包，发送时把时间、IP 添加到一个超时缓存里面，检查返回包 IP 是否在缓存中，如果是再检查是否成功的协商包。完毕。

**Comment by SeaHOH** [comment]
如果要判断证书，可以把上面得到的 IP 用 client 获取证书。

**Comment by Kisesy** [comment]
@MeABc 好吧，其实说实话，我根本就没看过 quic-go 的代码，我一开始写代码就是照着例子写的，发现这样比较慢，想改成你说的那样，发现 quic-go 设置超时的时候设置不上的时候，我就知道事情没那么简单
我就放弃了 :trollface:

**Comment by Kisesy** [comment]
@SeaHOH 行我再看看吧，我先再部署个网页编译的app

**Comment by tec1987** [comment]
刚才试了一下，要不就是扫不到IP，要不就是卡住没反应(抓包没有数据)
XP系统，不知道是不是bug，我下载的是首次长传的编译好的程序 gscan_quic_windows_386.zip
![image](https://user-images.githubusercontent.com/23297763/28241294-62bc1e06-69c4-11e7-9e57-74b0b3fc053f.png)

**Comment by Kisesy** [comment]
@tec1987 好吧，出现了很多这种问题，不过大部分人也使用正常，所以不太好判断... 以后看吧

**Comment by tec1987** [comment]
可否加个debug参数，在关键部分打印调试信息，方便问题定位。。。

**Comment by Kisesy** [comment]
@tec1987 你把线程数调的很小试试

**Comment by Kisesy** [comment]
@MeABc 好了，以后不再需要自己编译了，现在自动编译自动上传了，跟社长的效果一样，爽翻天

**Comment by tec1987** [comment]
@Kisesy 
这就是8线程时的情况。。。

**Comment by Kisesy** [comment]
@tec1987 下载最新版试试吧

**Comment by tec1987** [comment]
@Kisesy 
新版OK，目前测试正常

**Comment by Kisesy** [comment]
@tec1987 确实好了？

**Comment by tec1987** [comment]
是啊能扫到结果，也不卡住了

**Comment by Kisesy** [comment]
@tec1987 你可能没仔细看以前的帖子，我说过不能改 HTTPVerifyHosts 的
尤其是改成 www.google.com，因为它默认会跳到 www.google.com.hk，然后就是重定向，挺麻烦的
而且旧版的这个配置是写死在代码中的，所以改了对不上必出错。
这个并发不好说，只能自己看着调吧。

**Comment by wpyok500** [comment]
@Kisesy  问个题外话，如何实现exe+配置文件？

**Comment by Kisesy** [comment]
@wpyok500 这里不是有写吗 https://github.com/Kisesy/gscan_quic/releases

**Comment by Kisesy** [comment]
@Asteroids89 @tec1987  都来试试最新版吧，也许卡住不一定解决，连接数应该没问题了

**Comment by wpyok500** [comment]
@Kisesy 你这个是每次版本更新都要下载配置文件，还是exe文件，还是两个都要下载更新?

**Comment by Kisesy** [comment]
@wpyok500 有了配置就不需要再下了，把更新的编译文件放到你配置的文件夹里覆盖就行了

**Comment by tec1987** [comment]
@Kisesy 原来是这样，忘记重定向的问题了。。。

**Comment by qwerttaa** [comment]
我现在第二次用，线程800

第一次时候是500

下次用，我试试1000

路由是斐讯k3

**Comment by Asteroids89** [comment]
已下载，谢谢主任。

**Comment by smovie** [comment]
下载的1.2版本也是楼主的情况，系统是win7x64
然后拷到另一台机器win2003就可以正常扫描

**Comment by candy2107** [comment]
@phuslu
社长来这里借鉴点经验吧
